### ISSUE TITLE
No Sanitize applied on the 'Key' - XSS

### ISSUE THREAD

**Issue by carlo-feudo** [angular-translate/angular-translate#1418]
Hi there, doing some tests I have seen that it could occur a pretty nasty issue.
I.g. when the value passed to the translate= directive is a script (XSS) it would bypass the sanitize provider.

Please have a look at this plunker http://plnkr.co/edit/jTMAtvHcwsJZLi4f355H?p=preview

```
app.controller('Ctrl',` ['$scope', function ($scope) {
$scope.a = '<script>alert(1)</script>';
}]);

issue --->

  <p translate="{{a}}" translate-values="{username: '<span style=\'color:red;\'>HACKED</span>'}"></p>

```

Cheers, 
Carlo

**Comment by tspaeth** [comment]
Hm, if I exactly call your plnkr, my browser puts

```
&lt;script&gt;alert(1)&lt;/script&gt;
```

into the output - isn't that expected and sanitizeD?

**Comment by carlo-feudo** [comment]
Hello, be aware that there are two line printed in the example. 

The first is correct and expected, the second one is hidden by plunked engine because there is actually the Script executable as dom node.

Cheers
Carlo

**Comment by tspaeth** [comment]
Now I see. yes. But it is not executed?

**Comment by carlo-feudo** [comment]
Not in plunker but in a normal app it is executed by the browser unfortunately, shall we extend the sanitation to the keys as well? 

Sounds pretty dangerous. I manage to execute jQuery code.

Cheers
Carlo

**Comment by carlo-feudo** [comment]
Hi there, 
the problem is the line number **2871** (angular-translate.js, directive translate) as for the following:

`iElement.empty().append(scope.preText + value + scope.postText);
`
basically the function jQuery **append** will append the text (key) as node html.

If you modify the line with:

`iElement.empty().**append**(scope.preText + $translateSanitization.sanitize(value) + scope.postText);
`
it still wont work.

To make it "work" apparently we need to replace the **append** method with the **text** method, like

`iElement.empty().**text**(scope.preText + $translateSanitization.sanitize(value) + scope.postText);
`

Thoughts?

Cheers, 
Carlo

**Comment by knalli** [comment]
> Not in plunker but in a normal app it is executed 

Again, please provide a working demo demonstrating the issue. How should we very this will work?

**Comment by carlo-feudo** [comment]
Hi, as you can see from the chrome inspector the script attached form the directive translate is a real DOM node as the function sanitize function is not in place for keys and with the append() method it will be transformed in a real node.

 
<img width="1118" alt="screen shot 2016-02-26 at 17 25 32" src="https://cloud.githubusercontent.com/assets/9084808/13359908/582dbe12-dcae-11e5-86cf-94e5df2c2861.png">

**Comment by knalli** [comment]
Sorry for the delay, I had not complete my inspections on this issue days ago.

At the moment I'm wondering what the implication is exactly. The demo is demonstrating nothing is executed actually, but I see the general problem. However, this means there are two requirements:

1. the key is malicious (in case of the key is computed with potential malicious user input)
2. the key is the inner content

Anyway: We need a working demo exploiting things. Otherwise we can't say we would fix this.

**Comment by carlo-feudo** [comment]
Hi @knalli I have mange to reproduce a nasty one.

I want to say that sometimes we could use angular-translate to handle keys coming from the server, hence this could be nasty.

I have published the test on Heroku, have a look:  https://angular-translate-test.herokuapp.com/

you will se that the script inject as key is:

`  this.a = "<div onmouseover=\"javascript:alert(2);\">MOUSE OVER ME!</div>";
`

Please let us know what you think

Cheers, 
Carlo

**Comment by knalli** [comment]
Okay, now we have a working demo.

http://plnkr.co/edit/KJVdvDNI8P2LEOK4uFAi?p=preview

**Comment by tspaeth** [comment]
So at the end in this case it means:

If a translation key has no corresponding value in any language and the default behavior to show the key again is used - it might open a kind of XSS issue.

**Comment by carlo-feudo** [comment]
Definitely, a solution could be in case of key reflection to use the method .text() instead of .append()

Cheers

**Comment by blaryjp** [comment]
Hello guys,

I share my solution:  use missingTranslationHandlerLog

http://plnkr.co/edit/0BguvH79prDTvy0oJxE3?p=preview

Cheers,

**Comment by barillax** [comment]
Just a note, we saw this XSS in the wild in our app. We got lucky because the injected text was only displayed back to the user who had just entered it, but it basically followed the same formula described above:

1. User changed their `nickname` to be `><img src=x onerror=alert(document.domain)>`
2. App displays the user's nickname in a translated string like so:
`<p translate="welcome_x" translate-values="{'name', currentUser.nickname}"></p>`
3. Result: an alert box pops up.

**Comment by knalli** [comment]
Please provide a complete running example. Thanks.

**Comment by barillax** [comment]
My bad guys. I think I figured out that we had a codepath that was disabling sanitization in this particular scenario. Angular-translate seems to prevent the issue when "escape" is enabled as the sanitization strategy.

**Comment by knalli** [comment]
Ok, back to the actual issue.

1. Should we invest time in effort making the key XSS safe? It is definitely not recommended to use such keys, neither explicitly nor implicitly like based directly on user input. However, it is possible at the moment.
2. What is proper solution? [This](https://github.com/angular-translate/angular-translate/issues/1418#issuecomment-189268953) is not a solution, because HTML injection *is* a feature.

Sanitize/escape the requesting key if using as a default? Instead of applying `translationId` => `escape(translationId)` [here](https://github.com/angular-translate/angular-translate/blob/2.12.1/src/directive/translate.js#L274-L278)?

**Comment by bryanbraun** [comment]
@knalli, we ran into this issue recently.

In our case, it's hard to say "don't use translation based directly on user input," because many of our use-cases are for generic, reusable components, in a UI kit. We want the components to support translation, but a downstream user may need to display user content with them.

I feel like a good solution would be to sanitize the key with the same sanitization strategy that is set for the library as a whole.
